modules/control/submodules/preprocessor_submodule.cc

```diff
@@ -54,107 +55,55 @@ bool PreprocessorSubmodule::Init() {
-  CHECK(trajectory_reader_ != nullptr);
-
-  cyber::ReaderConfig localization_reader_config;
-  localization_reader_config.channel_name = FLAGS_localization_topic;
-  localization_reader_config.pending_queue_size =
-      FLAGS_localization_pending_queue_size;
-  localization_reader_ = node_->CreateReader<LocalizationEstimate>(
-      localization_reader_config, nullptr);
-  CHECK(localization_reader_ != nullptr);
-
-  cyber::ReaderConfig pad_msg_reader_config;
-  pad_msg_reader_config.channel_name = FLAGS_pad_topic;
-  pad_msg_reader_config.pending_queue_size = FLAGS_pad_msg_pending_queue_size;
-  pad_msg_reader_ =
-      node_->CreateReader<PadMessage>(pad_msg_reader_config, nullptr);
-  CHECK(pad_msg_reader_ != nullptr);
-
   // Preprocessor writer
   preprocessor_writer_ =
       node_->CreateWriter<Preprocessor>(FLAGS_control_preprocessor_topic);
-  CHECK(preprocessor_writer_ != nullptr);
 
-  // set initial vehicle state by cmd
-  // need to sleep, because advertised channel is not ready immediately
-  // simple test shows a short delay of 80 ms or so
-  AINFO << "Control resetting vehicle state, sleeping for 1000 ms ...";
-  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
-
-  // should init_vehicle first, let car enter work status, then use status msg
-  // trigger control
-  AINFO << "Control default driving action is: "
-        << DrivingAction_Name(control_common_conf_.action());
-  pad_msg_.set_action(control_common_conf_.action());
+  CHECK(preprocessor_writer_ != nullptr);
 
   return true;
 }
 
-bool PreprocessorSubmodule::Proc() {
-  chassis_reader_->Observe();
-  const auto chassis_msg = chassis_reader_->GetLatestObserved();
-  if (chassis_msg == nullptr) {
-    AERROR << "Chassis msg is not ready!";
-    return false;
-  }
-  OnChassis(chassis_msg);
-
-  trajectory_reader_->Observe();
-  const auto trajectory_msg = trajectory_reader_->GetLatestObserved();
-  if (trajectory_msg == nullptr) {
-    AERROR << "planning msg is not ready!";
-    return false;
-  }
-  OnPlanning(trajectory_msg);
-
-  localization_reader_->Observe();
-  const auto localization_msg = localization_reader_->GetLatestObserved();
-  if (localization_msg == nullptr) {
-    AERROR << "localization msg is not ready!";
-    return false;
-  }
-  OnLocalization(localization_msg);
-
-  pad_msg_reader_->Observe();
-  const auto pad_msg = pad_msg_reader_->GetLatestObserved();
-  if (pad_msg != nullptr) {
-    OnPad(pad_msg);
+bool PreprocessorSubmodule::Proc(
+    const std::shared_ptr<LocalView> &local_view,
+    const std::shared_ptr<PadMessage> &pad_message) {
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    local_view_reader_->Observe();
+    const auto &local_view_msg = local_view_reader_->GetLatestObserved();
+    if (local_view_msg == nullptr) {
+      AERROR << "local view msg is not ready!";
+      return false;
+    }
+    local_view_->CopyFrom(*local_view_msg);
```

Comment:
Add TODO here to try to duplicated copies.

Reply:
Added. Thank you.